Docs/PinterestFocusedXcodeProjects.md

```diff
@@ -0,0 +1,121 @@
+
+When changes land into master, Xcode projects are built, and developer can
+cache hit the project. This helps make large Xcode projects generate
+fast.
+
+An original limitation with project generation was knowing "when" to update the
+project. Originally, the developer was required to run xchammer after git
+checkouts which led to broken indexing and builds. Later, support was added to
+automatically update Xcode projects, which included a complex _and_slow_
+algorithm. By utilizing Bazel cache facilities and incremental build abilities
+we cut nooped generation time down from 3s to a couple hundred milliseconds.
+
+XCHammer will retain support of the original design as it allows people to try
+out XCHammer quite easily and is good for simple apps.
+
+### Xcode Project Configuration
+
+Originally, we stored configuration information in a yml representation of
+XCHammerConfig. This became hard to manage, and was [refactored to a skylark
+DSL](https://github.com/pinterest/xchammer/pull/133)
+to make it easier to specify several options.
+
+
+An example of this is our build instrumentation statsd reporting system, which
+is tacked on internally. It ties into schemes by adding `execution_actions` to
+run pre/post build
+
+```
+log_build_start_action = execution_action(
+    name = "Track build start",
+    script = "tools/instrumentation_helpers/log_build_start.py",
+)
+```
+
+Longer term we realized this wouldn't scale for the large number of diverse
+testing types. Xcode target settings are composed onto the build graph and
+specified adjacent to the Xcode target, mainly through custom macros. XCHammer
+aggregates all Xcode configuration for included targets [via an
+aspect](https://github.com/pinterest/xchammer/pull/192) which makes Xcode
+configuration composeable. By convention, the DSL is declared via macros
+adjacent to related BUILD files, e.g. for an an app declared at
+`Pinterest/iOS/App`, the Xcode configuration is declared at
+`Pinterest/iOS/App/XcodeProjectConfig.bzl`. The main BUILD file imports all of
+the configuration macros.
+
+### "XCFocus" aka focused projects
+
+Pinterest has a lot of source files which makes building, project generation,
+and indexing slow. We use Bazel locally to maximize caching and paralleism of
+clang invocations.
+
+Within the `XCHammerConfig` directory, users declare Xcode projects via the `xcode_project` rule in a way that's ignored from SCM. By default we focus by target and path to succinctly filter a subset of # transitive dependencies for specified targets.
+
+```
+# XCHammerConfig/Focux.bzl
+ pinterest_xcode_project(
+        targets = [ ":PinterestDevelopment" ],
+        paths = [ "Pinterest/**" ],
+    )
+```
+
+After editing the focused file, the Xcode project automatically is updated (
+add the this for Bazel builds )
```

Comment:
Incomplete thought?

Reply:
This is a TODO since we don't yet build Xcode projects for the Bazel targets

Reply:
https://github.com/pinterest/xchammer/issues/206